src/vs/platform/contextkey/common/contextkey.ts

```diff
@@ -129,6 +131,32 @@ export abstract class ContextKeyExpr {
 			return undefined;
 		}
 
+		const parser = new Parser();
+		const expr = parser.parse(serialized);
+		if (expr === undefined) { // print a console warning
+			const warning = [];
+			warning.push(`Failed to parse context key expression ("when clause"): ${serialized}\n`);
+			if (parser.lexingErrors.length > 0) {
+				warning.push('Lexing errors:\n\n');
+				parser.lexingErrors.forEach(token => warning.push(Scanner.reportError(token), '\n'));
+			}
+
+			if (parser.parsingErrors.length > 0) {
+				if (parser.lexingErrors.length > 0) { warning.push('\n\n'); } // separate lexing errors from parsing errors
+				warning.push('Parsing errors:\n\n');
+				parser.parsingErrors.forEach(message => warning.push(`${message}`, '\n'));
+			}
+			console.warn(warning.join(''));
```

Comment:
We shouldn't do console.warn here, this is too low level. I suggest we create a new variant of `ContextKeyExpr.deserialize`, perhaps `ContextKeyExpr.deserializeWithErrors` which would return the errors and then adopt that in the various callers. Those callers might have a chance to report more useful information, like for example the extension id that has the problematic context key, or the keybinding item, etc.